auctions/dutch-auction/src/main.sw

```
@@ -0,0 +1,204 @@
+
+        require(!auction.ended, TimeError::AuctionAlreadyEnded);
+
+        // Cannot bid before auction starts
+        require(auction.start_time <= height(), TimeError::AuctionNotYetStarted);
+        // Cannot bid after auction ends
+        require(height() <= auction.end_time, TimeError::AuctionAlreadyEnded);
+
+        // Checks for correct asset_id being sent
+        require(msg_asset_id() == auction.asset_id, UserError::WrongAssetSent);
+
+        let price = calculate_price(auction);
+
+        // Checks for high enough amount being sent
+        require(price <= msg_amount(), UserError::BidTooLow);
+
+        // Disallows furthur bids
+        auction.ended = true;
+        auction.winner = Option::Some(sender_indentity());
+        storage.auctions.insert(auction_id, auction);
+
+        // If someone sends more than the current price, refunds the extra amount
+        if msg_amount() > price {
+            let return_amount = msg_amount() - price;
+            transfer_to_identity(return_amount, auction.asset_id, sender_indentity());
+        }
+
+        on_win(auction, price);
+
+        let _ = storage.active_auctions_of_author.pop(auction.author);
+
+        storage.auctions_won.push(sender_indentity(), auction_id);
+
+        log(WinningBidEvent {
+            id: auction_id,
+            winner: sender_indentity(),
+        });
+    }
+
+    #[storage(read, write)]
+    fn create_auction(
+        opening_price: u64,
+        reserve_price: u64,
+        start_time: u64,
+        end_time: u64,
+        beneficiary: Identity,
+        asset: ContractId,
+    ) {
+        require(reserve_price <= opening_price, SetupError::EndPriceCannotBeLargerThanStartPrice);
+        require(height() < end_time, SetupError::AuctionCannotEndInThePast);
+        require(height() <= start_time, SetupError::AuctionCannotStartInThePast);
+        require(start_time < end_time, SetupError::AuctionCannotEndBeforeItStarts);
+
+        let auction = Auction {
```

I'd consider adding a `new()` method.

---

I dont see a point, every field needs to be set either way

---

1. There are 2 default values at the end that can be hardcoded and obscured
2. It makes the main file easier to read on a higher level since it's cleaner